### PR TITLE
adding ALS and LogisticFM, and allow sigmoid and raw activatoin for lowrank df

### PR DESCRIPTION
**Pull request by lgsaber**
adding ALS and LogisticFM, and allow sigmoid and raw activatoin for lowrank df

*Description of changes: 
1) added two model classes under models/implicit.py, which are based on ALS and LogisticMF implementation from Implicit. 
2) extend ExponentiatedLowRankDataFrame to CustomLowRankDataFrame to allow additoinal activation functions (sigmoid, raw logit).
3) modified environmental configuration to add Implicit.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/rim_experiments/util/score_array.py [hunk 1]
```diff
@@ -104,6 +104,32 @@ def collate_fn(cls, batch):
...
+class SigmoidLowRankValues(ExponentiatedLowRankValues):
+    """ mimics a numpy 2-d array with sigmoid low-rank structures
+    """
+    def eval(self, device=None):
+        if device is None:
+            return 1/(1-np.exp(self.ind_logits @ self.col_logits.T))
+        else:
+            ind_logits = torch.as_tensor(self.ind_logits, device=device)
+            col_logits = torch.as_tensor(self.col_logits, device=device)
+            return (ind_logits @ col_logits.T).sigmoid()
```
Comment (yifeim) [inline src/rim_experiments/util/score_array.py:117]: `* self.sign` because this may be multiplied with scalars in cvx.

File: src/rim_experiments/util/score_array.py [hunk 2]
```diff
@@ -104,6 +104,32 @@ def collate_fn(cls, batch):
...
+class DotProdLowRankValues(ExponentiatedLowRankValues):
+    """ mimics a numpy 2-d array with raw dot product low-rank structures
+    """
+    def eval(self, device=None):
+        if device is None:
+            return self.ind_logits @ self.col_logits.T
+        else:
+            ind_logits = torch.as_tensor(self.ind_logits, device=device)
+            col_logits = torch.as_tensor(self.col_logits, device=device)
+            return (ind_logits @ col_logits.T)
```
Comment (yifeim) [inline src/rim_experiments/util/score_array.py:130]: `* self.sign` because it may be multiplied with scalars in cvx.

File: src/rim_experiments/util/score_array.py [hunk 3]
```diff
@@ -176,3 +202,84 @@ def __mul__(self, other):
...
+
+
+    def reindex(self, index, axis=0, fill_value=float("nan")):
+        if axis==1:
+            return self.T.reindex(index, fill_value=fill_value).T
+
+        ind_logits = np.pad(self.ind_logits, ((0,1), (0,1)), constant_values=0)
+        #TODO: change fillna according to activation type
+        with np.errstate(divide='ignore'): # 0 -> -inf
+            ind_logits[-1, -1] = np.log(fill_value)
```
Comment (yifeim) [inline src/rim_experiments/util/score_array.py:261]: `np.log(fill_value) if self.activation in ['exponential', 'sigmoid'] else fill_value # raw(0)`

File: src/rim_experiments/util/score_array.py [hunk 4]
```diff
@@ -176,3 +202,84 @@ def __mul__(self, other):
...
+            np.arange(len(self)), index=self.index
+            ).reindex(index, fill_value=-1).values
+
+        return self.__class__(
+            ind_logits[new_ind], col_logits, self.sign, index, self.columns,self.activation)
+
+
+    def __mul__(self, other):
+        """ left-join multiply with other dataframes; fill_value=0 on missing """
+        if isinstance(other, self.__class__):
```
Comment (yifeim) [inline src/rim_experiments/util/score_array.py:274]: `and other.activation == 'exponential'`?
